### PR TITLE
Added ComplexKeyRangeHashed dictionary

### LINKED ISSUE #22029: ComplexKey support for RangeHashedDictionary
Add support for complex key into range hashed dictionary.

### PR DESCRIPTION
**Pull request by kitaisreal**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- New Feature

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added ComplexKeyRangeHashed dictionary. Closes #22029.

### CODE REVIEW

File: src/Dictionaries/RangeDictionarySource.h [hunk 1]
```diff
@@ -14,170 +14,241 @@
 namespace DB
 {
 
-template <typename RangeType>
+enum class RangeDictionaryType
+{
+    simple,
```
Comment (akuzm) [inline src/Dictionaries/RangeDictionarySource.h:19]: Should start with a capital according to our code style. Same for another enum above.

**Review by akuzm — APPROVED** [review]
Makes sense.

File: src/Dictionaries/RangeDictionarySource.h (hunk 1 above)
Comment (kitaisreal) [reply src/Dictionaries/RangeDictionarySource.h:19]: @akuzm agreed, will fix.

**Comment by sevirov** [comment]
Internal documentation ticket: DOCSUP-13329

**Pull request merged**